pkg/kubectl/apply/element.go

```diff
@@ -0,0 +1,375 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apply
+
+import (
+	"fmt"
+)
+
+// Element contains the record, local, and remote value for a field in an object
+// and metadata about the field read from openapi.
+// Calling Merge on an element will apply the passed in strategy to Element -
+// e.g. either replacing the whole element with the local copy or merging each
+// of the recorded, local and remote fields of the element.
+type Element interface {
+	// FieldMeta specifies which merge strategy to use for this element
+	FieldMeta
+
+	// Merge merges the recorded, local and remote values in the element using the Strategy
+	// provided as an argument.  Calls the type specific method on the Strategy - following the
+	// "Accept" method from the "Visitor" pattern.
+	// e.g. Merge on a ListElement will call Strategy.MergeList(self)
+	// Returns the Result of the merged elements
+	Merge(Strategy) (Result, error)
+
+	// HasRecorded returns true if the field was explicitly
+	// present in the recorded source.  This is to differentiate between
+	// undefined and set to null
+	HasRecorded() bool
+
+	// GetRecorded returns the field value from the recorded source of the object
+	GetRecorded() interface{}
+
+	// HasLocal returns true if the field was explicitly
+	// present in the recorded source.  This is to differentiate between
+	// undefined and set to null
+	HasLocal() bool
+
+	// GetLocal returns the field value from the local source of the object
+	GetLocal() interface{}
+
+	// HasRemote returns true if the field was explicitly
+	// present in the remote source.  This is to differentiate between
+	// undefined and set to null
+	HasRemote() bool
+
+	// GetRemote returns the field value from the remote source of the object
+	GetRemote() interface{}
+}
+
+// FieldMeta defines the strategy used to apply a Patch for an element
+type FieldMeta interface {
+	// GetFieldMergeType returns the type of merge strategy to use for this field
+	// maybe "merge", "replace" or "retainkeys"
+	// TODO: There maybe multiple strategies, so this may need to be a slice, map, or struct
+	// Address this in a follow up in the PR to introduce retainkeys strategy
+	GetFieldMergeType() string
+
+	// GetFieldMergeKeys returns the merge key to use when the MergeType is "merge" and underlying type is a list
+	GetFieldMergeKeys() MergeKeys
+
+	// GetFieldType returns the openapi field type - e.g. primitive, array, map, type, reference
+	GetFieldType() string
+}
+
+// HasElementData contains whether a field was set in the recorded, local and remote sources
```

Comment:
I find it surprising to bundle the `Has` property of these three objects like you did, i.e. 
```
type HasElementData struct {
...
}
```
Rather than create individual objects:
```
type ElementData interface {
    Has() bool
    Get() interface{}
}
```

though I understand that it might be difficult to manipulate/instantiate.

Reply:
I think we want to keep the recorded, local, and remote values together in a single data structure.  I could merge the HasElementData into the RawElementData so that has the `Has` functions.  WDYT?

Reply:
Yeah, probably. I think the `has` function should be close to the its data.

Reply:
I agree.